### DISCUSSION TITLE
Query regarding `QoSOptions` implemention in version 23.2+

### DISCUSSION THREAD

**Discussion by ankits23560** [ThreeMammals/Ocelot#2112] [Q&A] [answered]
Hello everyone,
 
We are currently experiencing difficulties while trying to implement QoS with Polly V8 in version 23.3.3.
The Issue we are facing is explained below:

Under Polly V7 version 23.2.2:
 
**JSON Configuration:**
```json
"QoSOptions": {
    "ExceptionsAllowedBeforeBreaking": 3, // Number of retries before circuit breaker opens
    "DurationOfBreak": 15000, // Duration in milliseconds the circuit breaker will be open, sending back a 503
    "TimeoutValue": 5000 // Duration in milliseconds waiting for a response before a timeout occurs
}
```
**Program. cs Configuration:**
```csharp
services.AddJwtAuthentication();
services.AddOcelot().AddPollyV7();
```
**Behavior Of QoSOptions with Polly V7:**
On Implementing QoSOptions with Polly V7 we observed that the circuit breaker was breaking the circuit when the Exceptions occurred more than 3 times and also breaking the circuit when the timeout exceptions exceeded 3 times. However, on implementing QoS with Polly v8 the circuit is breaking the circuit with the exceptions but not breaking the circuit when the timeout exceptions.
 

We would appreciate your help in generating the scenario mentioned above for implementing the QoS circuit breaker with Polly V8 in version 23.3.3 since it is not clearly explained in the Polly documentation.
 
Is it possible to implement the circuit breaker strategy and the timeout strategy together?
 
Is there any workaround regarding this issue?

Also, How can we implement the QoS behavior of Polly v7 in Polly v8?
Also Requesting can you please update the documentation for Ocelot regarding this issue.
Thanks
Ankit Sukhwani

**Comment by ankits23560** [comment]
hi @RaynaldM  can you please guide me on the same?

**Reply by raman-m** [reply]
In @RaynaldM  we trust! 😄

**Reply by ankits23560** [reply]
Hi @raman-m, could you please provide any inputs that I can try?

**Reply by raman-m** [reply]
@ankits23560 Read [my answer](https://github.com/ThreeMammals/Ocelot/discussions/2112#discussioncomment-9958030) below plz! 👇

**Comment by raman-m** [comment, marked as answer]
Hello Ankit!

>#### Behavior Of QoSOptions with Polly V7:

I'm not sure why you're researching the previous version 23.2 vs Polly v7. It would be better to migrate to the new version 23.3 vs Polly v8. Our team's intentions have been explained in the [Ocelot extra packages](https://ocelot.readthedocs.io/en/23.3.0/index.html#ocelot-extra-packages) section of the Release Notes.
Support for version 23.2 will be discontinued. Our release strategy focuses on developing new versions without maintaining older ones.

---
> On Implementing QoSOptions with Polly V7 we observed that the circuit breaker was breaking the circuit when the Exceptions occurred more than 3 times and also breaking the circuit when the timeout exceptions exceeded 3 times. **However, on implementing QoS with Polly v8 the circuit is breaking the circuit with the exceptions but not breaking the circuit when the timeout exceptions.**

It's unusual! We already manage timeout exceptions in the configuration of the Circuit Breaker strategy:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot.Provider.Polly/PollyQoSResiliencePipelineProvider.cs#L92-L93
Can you specify which exceptions you're referring to? What are the exact types?

Additionally, we log `Information` messages at this location:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot.Provider.Polly/PollyQoSResiliencePipelineProvider.cs#L130
Could you modify the logging settings to utilize the `Information` level to view the provider's messages?
Please examine your logs for specific messages and provide us with the complete request log.

----
> We would appreciate your help in generating the scenario mentioned above for implementing the QoS circuit breaker with Polly V8 in version 23.3.3 since it is not clearly explained in the Polly documentation.

The details are outlined in our [Quality of Service](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html) documentation. However, it seems you might have some customizations for the QoS feature, correct? Without seeing your solution, it's challenging to offer specific advice. If your scenario involves both strategies, please review the [Configuration](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html#configuration) section again.

I'm concerned that your configuration's `DurationOfBreak` is three times greater than the `TimeoutValue`, with 15000 being more than three times 5000. This configuration implies that most requests will time out when the Circuit is open. Moreover, if the Circuit is closed, the Circuit Breaker strategy will absorb timeout exceptions due to the default settings.

Should you require behavior that deviates from the default, you'll need to create your own provider utilizing the [Extensibility](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html#extensibility) feature.

---
>Is it possible to implement the circuit breaker strategy and the timeout strategy together?

Indeed, it is possible to implement both the circuit breaker and timeout strategies concurrently. For guidance, consult the [Configuration](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html#configuration) section in the documentation. Remember, you must define all three properties to activate both strategies.

----
> Is there any workaround regarding this issue?

To address this issue, I suggest the following:
- Set `DurationOfBreak` to be equal to or less than `TimeoutValue`. It is generally advised that `TimeoutValue` be significantly higher than `DurationOfBreak`.
- Activate the `Information` logging level and conduct further investigation. This lower level of logging should provide sufficient details to comprehend the behavior of your QoS configuration.
- Don't use `AddPollyV7` and version 23.2.x! Use `AddPolly` and version 23.3+!!!

----
> Also, How can we implement the QoS behavior of Polly v7 in Polly v8?

Seems you have to implement your own QoS provider using the [Extensibility](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html#extensibility) feature.
Please learn official services helpers C# code:
- Polly QoS [OcelotBuilderExtensions](https://github.com/ThreeMammals/Ocelot/blob/23.3.0/src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs#L16)

Keep in mind that we no longer support Polly v7❗  The Polly library released version 8 in the previous year, 2023. It makes no sense to develop using Polly v7, which is why we have removed the v7 logic and provider.

----
>Also Requesting can you **please update the documentation** for Ocelot regarding this issue.

Our official [documentation](https://ocelot.readthedocs.io/en/23.3.0/features/qualityofservice.html) already contains sufficient information and links to Polly for implementing QoS in Ocelot solutions.

Hope it helps!

**Reply by ankits23560** [reply]
Hi @raman-m ,
Apologies for the late reply. I was occupied with some other simultaneous development. I have implemented the changes according to your recommendation of setting the timeout value less than the duration of the break. However, I am still unable to comprehend the behavior of the resilience pipeline.

> I'm not sure why you're researching the previous version 23.2 vs Polly v7. It would be better to migrate to the new version 23.3 vs Polly v8. Our team's intentions have been explained in the [Ocelot extra packages](https://ocelot.readthedocs.io/en/23.3.0/index.html#ocelot-extra-packages) section of the Release Notes.
Support for version 23.2 will be discontinued. Our release strategy focuses on developing new versions without maintaining older ones.

It is for the comparision for circuit breaker in Polly v7 which is not working properly with Polly v8.
 The error scenario I encountered is outlined below:
**Configuration:**
```json
{
"UpstreamPathTemplate": "/api/ForQosTest",
"UpstreamHttpMethod": [ "GET" ],
"DownstreamPathTemplate": "/ForQosTestCase",
"DownstreamScheme": "http",
"DownstreamHttpMethod": "GET",
"DownstreamHostAndPorts": [
  {
    "Host": "Localhost", "Port": 3000
  }
],
"QoSOptions": 
{
  "ExceptionsAllowedBeforeBreaking": 3, //no of retries before circuit breaker opens
  "DurationOfBreak": 5000, //no of Milli seconds circuit breaker will be open sending 503 back
  "TimeoutValue": 15000 //no of Milli seconds waiting for response before timeout happens
},
```
> Can you specify which exceptions you're referring to? What are the exact types?

The error i am referring to is TimeOutstrategy. As suggested by you the circuit does not break the circuit when the configured api
is giving delayed response.

>Indeed, it is possible to implement both the circuit breaker and timeout strategies concurrently. For guidance, consult the Configuration section in the documentation. Remember, you must define all three properties to activate both strategies.

we have defined all the three configuration as mentioned in the documentation.

> Keep in mind that we no longer support Polly v7❗ The Polly library released version 8 in the previous year, 2023. It makes no sense to develop using Polly v7, which is why we have removed the v7 logic and provider.

Noted we have kept in my regarding the depletion of polly v7 so we have upgraded to polly v8 but there is one functionality which was implemented polly v7 but I guess its not working in polly v8 or might be I am doing some mistakes while configuring the QoSOptions. 

Also Can you please guide me on this ?
Or is anything that is missing while configuring the V8 Circuit breaker?
The intent is to know the exact behavior of the circuit breaker in polly v8...
Please do refer to the logs for the reference of the issue i am facing while configuring the polly V8 in v23.3+
[logs.txt](https://github.com/user-attachments/files/16284623/logs.txt)

**Reply by raman-m** [reply]
> Also Can you please guide me on this?

Apologies, Ankit! I'm unable to provide further consultation. Should you require the previous functionality, please continue using the older versions of Ocelot.

> Also, How can we implement the QoS behavior of Polly v7 in Polly v8?

To implement the QoS behavior of Polly v7 in Polly v8, it is advisable to consult the [Polly documentation](https://www.pollydocs.org/) and engage with the Polly community regarding the [migration guide from v7 to v8](https://www.pollydocs.org/migration-v8.html).

**Reply by raman-m** [reply]
@RaynaldM @ggnaegi 
In the [logs.txt](https://github.com/user-attachments/files/16284623/logs.txt), I discovered that the `TimeoutException` is initiated by the `Ocelot.Requester.TimeoutDelegatingHandler`. The author claims that this exception is not captured by the Circuit Breaker, resulting in the circuit not being broken by the first strategy in v23.3. However, we have configured the `TimeoutException` to be managed by the circuit breaker:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot.Provider.Polly/PollyQoSResiliencePipelineProvider.cs#L92-L93
I am unable to confirm or refute this user scenario and would appreciate your insights. Regrettably, replicating this user scenario necessitates a significant time investment to write an acceptance test to reproduce the issue.
Could you recommend something plz?

**Reply by ggnaegi** [reply]
@raman-m We should get rid of the Timeout Delegating Handler... I told you that a few months ago :-)

**Reply by raman-m** [reply]
@ggnaegi 
After removing that, will the QoS stabilize? It appears this is your design and logic. I apologize, but I don't recall our previous discussions on this matter. Could you provide any links? I only remember that we have discussed this.

Regarding the next question, the `TimeoutDelegatingHandler` class is pivotal to our timeout feature. If we eliminate this handler, what will be our method for configuring timeouts? Will it be based on `HttpRequestMessage`? Also, we no longer utilize `HttpClient` in the kernel.
As for now, I don't see an alternative for timeout feature.

Should we convert this discussion into an issue, or would you prefer to continue in this thread? Please let me know when you will be available to work on this.

**Reply by raman-m** [reply]
I apologize, @ggnaegi. Are you suggesting that the handler ought to be removed so that QoS routes can be managed by the TimeOut strategy, while the handler should be retained for regular non-QoS routes to be managed by `MessageInvokerPool`?

**Reply by ankits23560** [reply]
hi @ggnaegi it seems that this is an issue. We are facing major problems in implementing circuit breaker with timeout-strategy. 
Is there any workaround going on regarding this thread?

**Reply by raman-m** [reply]
@ankits23560 Dear Ankit,
It appears that our QoS logic in version 23.3 does not meet your needs.
I suggest disabling the [QoS](https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html) feature entirely and creating your own [delegating handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) where you can implement the previous Polly v7 logic or [migrate to v8](https://www.pollydocs.org/migration-v8.html) seamlessly. [Our team plans](https://github.com/ThreeMammals/Ocelot/discussions/2112#discussioncomment-10507241) to phase out the [TimeoutDelegatingHandler](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20TimeoutDelegatingHandler&type=code) in the near future, possibly in the forthcoming releases.

#### P.S.
Please review [note 1](https://ocelot.readthedocs.io/en/latest/features/qualityofservice.html#notes) and inform us if the exception has been resolved in your logs. Thank you for your understanding.

#### P.P.S.
Please be aware that we have an open PR #2073, which involves a redesign of the Timeout feature. This PR will have implications for Quality of Service (QoS).

**Comment by ankits23560** [comment]
Hi @raman-m,
 can we move this discussion to the issue? We still have no confirmation about the issue?

**Reply by raman-m** [reply]
Please display the issue in the open pull request. It does not matter whether you attach the PR to the discussion thread or to an issue thread.

**Reply by raman-m** [reply]
👉 https://github.com/ThreeMammals/Ocelot/discussions/2112#discussioncomment-10095108

**Comment by raman-m** [comment]
@ggnaegi wrote [on Jul 19, 2024](https://github.com/ThreeMammals/Ocelot/discussions/2112#discussioncomment-10095720)
>We should get rid of the Timeout Delegating Handler... I told you that a few months ago :-)

Alright, Gui, you told me [here](https://github.com/ThreeMammals/Ocelot/discussions/2072#discussioncomment-9469487)... I now have a clearer understanding of the timeout logic 😇 and it appears we can replace our custom [TimeoutDelegatingHandler](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20TimeoutDelegatingHandler&type=code) with the [SocketsHttpHandler class](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler?view=net-8.0). Does that sound good?
Regarding the delivery of #2073, should it be before or after the removal of `TimeoutDelegatingHandler`?
It seems we should deliver #2073 first, followed by the removal of `TimeoutDelegatingHandler`, as it requires significant refactoring of the core messaging logic, correct?

I propose the following task &rarr;
### TODO: Implement Timeout Logic Using the SocketsHttpHandler Class
We need to utilize the following [properties](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler?view=net-8.0#properties):
- [ConnectTimeout](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.connecttimeout?view=net-8.0)
- [Expect100ContinueTimeout](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.expect100continuetimeout?view=net-8.0)
- [KeepAlivePingTimeout](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.keepalivepingtimeout?view=net-8.0)
- [PooledConnectionIdleTimeout](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.pooledconnectionidletimeout?view=net-8.0)
- [ResponseDrainTimeout](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.socketshttphandler.responsedraintimeout?view=net-8.0)

**Reply by raman-m** [reply]
Added the task to [TODOs aka Follow-up tasks](https://github.com/ThreeMammals/Ocelot/issues/2141#issuecomment-2322927815)
